Mathlib/FieldTheory/SeparablyGenerated.lean

```diff
@@ -0,0 +1,319 @@
+public import Mathlib.RingTheory.Polynomial.GaussLemma
+public import Mathlib.Algebra.CharP.IntermediateField
+
+/-!
+
+# Separably generated extensions
+
+We aim to formalize the following result:
+
+Let `K/k` be a finitely generated field extension with characteristic `p > 0`, then TFAE
+1. `K/k` is separably generated
+2. If `{ sᵢ } ⊆ K` is an arbitrary `k`-linearly independent set,
+  `{ sᵢᵖ } ⊆ K` is also `k`-linearly independent
+3. `K ⊗ₖ k^{1/p}` is reduced
+4. `K` is geometrically reduced over `k`.
+5. `k` and `Kᵖ` are linearly disjoint over `kᵖ` in `K`.
+
+## Main result
+- `exists_isTranscendenceBasis_and_isSeparable_of_linearIndepOn_pow`: (2) ⇒ (1)
+
+-/
+
+noncomputable section
+
+section
+
+attribute [local instance 2000] Polynomial.isScalarTower Algebra.toSMul IsScalarTower.right
+
+open MvPolynomial
+open scoped IntermediateField
+
+variable {k K ι : Type*} [Field k] [Field K] [Algebra k K] (p : ℕ) (hp : p.Prime)
+variable {a : ι → K} (n : ι) (ha : IntermediateField.adjoin k (Set.range a) = ⊤)
+variable (ha' : IsTranscendenceBasis k fun i : {i // i ≠ n} ↦ a i)
+variable (H : ∀ s : Finset K,
+  LinearIndepOn k _root_.id (s : Set K) → LinearIndepOn k (· ^ p) (s : Set K))
+
+namespace MvPolynomial
+
+/-- View a multivariate polynomial `F(x₁,...,xₙ)` as a polynomial in `xᵢ` with coefficients
+in `F(x₁,...,xᵢ₋₁,xᵢ₊₁,...,xₙ)`. -/
+def toPolynomialAdjoinImageCompl (F : MvPolynomial ι k) (a : ι → K) (i : ι) :
+    Polynomial (Algebra.adjoin k (a '' {i}ᶜ)) :=
+  letI := Classical.typeDecidableEq ι
+  (optionEquivLeft k _ (renameEquiv k (Equiv.optionSubtypeNe i).symm F)).mapAlgHom
+    (aeval fun j : {j // j ≠ i} ↦
+      (⟨a j, Algebra.subset_adjoin ⟨j, j.2, rfl⟩⟩ : Algebra.adjoin k (a '' {i}ᶜ)))
+
+theorem aeval_toPolynomialAdjoinImageCompl_eq_zero
+    {a : ι → K} {F : MvPolynomial ι k} (hFa : F.aeval a = 0) (i : ι) :
+    (toPolynomialAdjoinImageCompl F a i).aeval (a i) = 0 := by
+  rw [← hFa, ← AlgHom.restrictScalars_apply k]
+  simp_rw [toPolynomialAdjoinImageCompl, ← AlgEquiv.coe_algHom, ← AlgHom.comp_apply]
+  congr; ext; aesop (add simp optionEquivLeft_X_some) (add simp optionEquivLeft_X_none)
+
+theorem irreducible_toPolynomialAdjoinImageCompl {F : MvPolynomial ι k} (hF : Irreducible F) (i : ι)
+    (H : AlgebraicIndependent k fun x : {j | j ≠ i} ↦ a x) :
+    Irreducible (toPolynomialAdjoinImageCompl F a i) := by
+  have : a '' {i}ᶜ = Set.range (fun x : {j | j ≠ i} ↦ a x) := by aesop
+  delta toPolynomialAdjoinImageCompl
+  convert hF.map (renameEquiv k (Equiv.optionSubtypeNe i).symm) |>.map (optionEquivLeft k _) |>.map
+    (Polynomial.mapAlgEquiv (H.aevalEquiv.trans
+      (Subalgebra.equivOfEq _ _ congr(Algebra.adjoin k $this.symm))))
+  rw [← AlgEquiv.coe_algHom]
+  congr
+  aesop
+
+variable {F : MvPolynomial ι k} (hF0 : F ≠ 0) (hFa : F.aeval a = 0)
+variable (HF : ∀ F' : MvPolynomial ι k, F' ≠ 0 → F'.aeval a = 0 → F.totalDegree ≤ F'.totalDegree)
+
+include hF0 hFa HF in
```

Comment:
I personally feel like this `include .. in` pattern hampers readability, and would prefer to just repeat the assumptions in the statements.